clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3208,6 +3212,18 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
   using SyclKernelFieldHandler::handleSyclSamplerType;
 };
 
+// Kernels are only the unnamed-lambda feature if the feature is enabled, AND
+// the first template argument has been corrected by the library to match the
+// functor type.
+static bool IsSYCLUnnamedKernel(Sema &SemaRef, const FunctionDecl *FD) {
+  if (!SemaRef.getLangOpts().SYCLUnnamedLambda)
+    return false;
+  QualType FunctorTy = GetSYCLKernelObjectType(FD);
+  QualType TmplArgTy =
+      calculateKernelNameType(SemaRef.Context, FD).getUnqualifiedType();
```

Comment:
It's a little bit odd that one of these functions returns an unqualified type while the other one returns a possibly qualified type. Should `calculateKernelNameType()` always return an unqualified type instead? Or are there times when that qualification is important for the kernel name?

Reply:
Also, can you double-check that we have tests for qualified kernel names and that we get the proper instantiations out of them?

Reply:
I think actually the unqualification here is wrong, we only want it in unqualified mode since we want to differentiate between:

`kernel<Foo>(...);`
and 
`kernel<const Foo>(...);`

So I'll remove the unqualification here.  As far as tests, I don't see any CFE tests, but I'm sure there has got to be one in at least one of the test suites SOMEWHERE.

Reply:
We've thought that in the past :-D Can you make sure we have some test coverage for the case you mention? That seems like something we'd want to explicitly document as not naming the same kernel.